### PR TITLE
1/3 Implement improved API implementation for database

### PR DESCRIPTION
**Pull request by rkuris**
One step along the way, broken up for easier review. Changes include:
 - Implements api::DbView for DbRev
 - The API currently returns references; this is challenging with the way shale works, so we're (temporarily) going to return owned values from some APIs

### CODE REVIEW

**Review by exdx — APPROVED** [review]

File: firewood/src/db.rs [hunk 1]
```diff
@@ -272,6 +273,43 @@ pub struct DbRev<S> {
...
+            .map_err(|e| api::Error::IO(std::io::Error::new(ErrorKind::Other, e)))
+    }
+
+    async fn val<K: api::KeyType>(&self, key: K) -> Result<Option<Vec<u8>>, api::Error> {
+        let obj_ref = self.merkle.get(key, self.header.kv_root);
+        match obj_ref {
+            Err(e) => Err(api::Error::IO(std::io::Error::new(ErrorKind::Other, e))),
+            Ok(obj) => match obj {
+                None => Ok(None),
+                Some(inner) => Ok(Some(inner.as_ref().to_owned())),
```
Comment (xinifinity) [inline firewood/src/db.rs:291]: nit: Naming is hard.., we discussed to no longer use `inner` as it didn't bring any clarification?

File: firewood/src/db.rs [hunk 2]
```diff
@@ -272,6 +273,43 @@ pub struct DbRev<S> {
...
+            },
+        }
+    }
+
+    async fn single_key_proof<K: api::KeyType, N: AsRef<[u8]> + Send>(
+        &self,
+        _key: K,
+    ) -> Result<Option<Proof<N>>, api::Error> {
+        todo!()
+    }
```
Comment (xinifinity) [inline firewood/src/db.rs:301]: If these are `todo` maybe we can do it in a follow up with real implementations?

File: firewood/src/db.rs (hunk 1 above)
Comment (richardpringle) [reply firewood/src/db.rs:291]: I think we need to stop using the `inner` pattern. Naming something `inner` when we've simply added a wrapper type is fine IMO

File: firewood/src/db.rs [hunk 3]
```diff
@@ -272,6 +273,43 @@ pub struct DbRev<S> {
...
+#[async_trait]
+impl<S: ShaleStore<Node> + Send + Sync> api::DbView for DbRev<S> {
+    async fn root_hash(&self) -> Result<api::HashKey, api::Error> {
+        self.merkle
+            .root_hash(self.header.kv_root)
+            .map(|h| *h)
+            .map_err(|e| api::Error::IO(std::io::Error::new(ErrorKind::Other, e)))
+    }
+
+    async fn val<K: api::KeyType>(&self, key: K) -> Result<Option<Vec<u8>>, api::Error> {
```
Comment (richardpringle) [inline firewood/src/db.rs:285]: It's infuriating that you *had* to change this.

**Review by richardpringle — APPROVED** [review]

File: firewood/src/db.rs (hunk 1 above)
Comment (rkuris) [reply firewood/src/db.rs:291]: Yeah. This is one of those sweep tasks we can assign as a good first issue. Anyone want to create one?

File: firewood/src/db.rs (hunk 2 above)
Comment (rkuris) [reply firewood/src/db.rs:301]: That's why this is 1/3 :)

File: firewood/src/db.rs (hunk 1 above)
Comment (richardpringle) [reply firewood/src/db.rs:291]: I don't think it's trivial to unravel all the ` SomeType { inner: SomeInnerType }` especially with all the misused trait-bounds

**Pull request merged**